torchao/prototype/intx/Intx.py

```diff
@@ -0,0 +1,210 @@
+      pack_dim: (int) dimension to pack along
+    """
+    bits_to_shard = {
+        1: ["int1_shard"],
+        2: ["int2_shard"],
+        3: ["int2_shard", "int1_shard"],
+        4: ["int4_shard"],
+        5: ["int4_shard", "int1_shard"],
+        6: ["int4_shard", "int2_shard"],
+        7: ["int4_shard", "int2_shard", "int1_shard"],
+    }
+    def __new__(
+        cls,
+        shards: List[torch.Tensor],
+        packed_shape: List[int],
+        bit_size: int,
+        pack_dim: int = -1,
+    ):
+        kwargs = {"device": shards[0].device}
+        kwargs["device"] = shards[0].device
+        kwargs["layout"] = shards[0].layout
+        kwargs["requires_grad"] = False
+        kwargs["dtype"] = torch.uint8
+        return torch.Tensor._make_wrapper_subclass(cls, packed_shape, **kwargs) 
+
+    def __init__(
+        self,
+        shards: List[torch.Tensor],
+        packed_shape: List[int],
+        bit_size: int,
+        pack_dim: int = -1,
+    ):
+        shards = [shard.to(torch.uint8) for shard in shards]
+        self.shard = shards
+        for i, atrib in enumerate(self.bits_to_shard[bit_size]):
+            setattr(self, atrib, shards[i])
+            
+        self.packed_shape = packed_shape
+        self.bit_size = bit_size    
+        self.pack_dim = pack_dim
+    
+    def get_shards(self):
+        return [getattr(self,i) for i in self.__class__.bits_to_shard[self.bit_size]]
+    
+    def __repr__(self):
+        return f"Int{self.bit_size}Tensor(shape = {self.packed_shape}, data = {unpack(self.get_shards(), self.bit_size, dim = self.pack_dim)})"
+    
+    def __tensor_flatten__(self):
+        
+        return self.__class__.bits_to_shard[self.bit_size], [self.packed_shape, self.bit_size, self.pack_dim]
+    @classmethod
+    def __tensor_unflatten__(
+        cls, tensor_data_dict, tensor_attributes, outer_size, outer_stride
+    ):
+        shards =  list(tensor_data_dict.values())
+        packed_shape, bit_size, pack_dim = tensor_attributes
+        return cls(shards, packed_shape, bit_size, pack_dim)
+
+    implements = classmethod(_implements)
+    __torch_dispatch__ = classmethod(_dispatch__torch_dispatch__)
+    __torch_function__ = classmethod(_dispatch__torch_function__)
+
+    def get_plain(self):
+        return unpack(self.get_shards(), self.bit_size, dim = self.pack_dim)
+    
+    # temporary until kernels on packed tensors are created
+    def apply_transformation(self, fn):
+        og = self.get_plain()
+        new = fn(og)
+        return self.from_uint8(new, self.bit_size, self.pack_dim)
+    
+    # temporary until kernels on packed tensors are created
+    def apply_fn_to_shards(self, fn):
+        new_shards = [fn(shard) for shard in self.get_shards()]
+        return self.__class__(new_shards, self.packed_shape, self.bit_size, self.pack_dim)
+    
+    @classmethod
+    def from_uint8(cls, int_data: torch.Tensor, bit_size, pack_dim: int = -1):
```

Comment:
nit: are we defining uintx here or intx? I think we probably need both, just want to make sure we are using the correct one here

Reply:
I focused on supporting unsigned ints because the bit packing is simpler. Maybe intx is a misleading name. I think adding integer support can come when adding better bitpack kernels

Reply:
OK, then renaming it to uintx would be better I think